SIMSBIOHUB 664:  Sample Periods CSV Import

MacDeluca opened the pull request:
## Links to Jira Tickets

- https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-664

## Description of Changes

- New Sample Period CSV import
- New DualImportButton - For workflows that need to import both individually and bulk (Critter / Sample Periods)

## Testing Notes

- Should be able to import a sample period CSV (you can download the template via the dialog)
- Note: This will not check for duplicates OR overlapping period start / end dates (as Macgregor suggested in the effort of getting this into Prod quickly)

MacDeluca committed:
feat: created sampling periods csv import service

MacDeluca committed:
feat: added new UI components for dual import button

MacDeluca committed:
feat: created sample period import service and endpoint

MacDeluca committed:
feat: added template and linked frontend to backend

MacDeluca committed:
'

MacDeluca committed:
Merge branch 'dev' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-664-sampling-periods-csv

MacDeluca committed:
fix: added tests for sample-period/import endpoint

MacDeluca committed:
fix: updated error and solution messages

MacDeluca committed:
feat: updated docs + improved props for DualImportButton

MacDeluca committed:
fix: merge conflict with dev resolved

mauberti-bc commented:
A few changes to add:
* Validate that the end date is after the start date
* Add constraint preventing overlapping periods for a (site, technique) combo in the database

mauberti-bc committed:
change dual button to separate add and import buttons

MacDeluca committed:
feat: added getStartDateIsBeforeEndDateRowValidator

MacDeluca committed:
chore: started the database migration to add the new constraint

MacDeluca committed:
Merge branch 'SIMSBIOHUB-664-sampling-periods-csv' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-664-sampling-periods-csv

MacDeluca committed:
feat: added trigger to detect overlapping sample periods

MacDeluca committed:
fix: updated documentation / replace Map with CaseInsensitiveMap

NickPhura commented on api/src/services/import-services/sampling-periods/import-sample-periods-service.ts:
```diff
@@ -0,0 +1,144 @@
+    const initialConfig: CSVConfig<SamplePeriodCSVStaticHeader> = {
+      staticHeadersConfig: {
+        SAMPLE_SITE: { aliases: ['SAMPLE SITE', 'SAMPLING_SITE', 'SAMPLING SITE', 'SITE', 'LOCATION', 'STATION'] },
+        TECHNIQUE_NAME: { aliases: ['TECHNIQUE NAME', 'METHOD_TECHNIQUE', 'METHOD TECHNIQUE', 'TECHNIQUE', 'METHOD'] },
```
nit: The name of this field is different from the observations field, and the array of aliases is also different.

I'm sure it doesn't change the functionality, but might be good to unify them if we can.

NickPhura replied:
Same with 'SAMPLING_SITE' vs 'SAMPLING_SITE', etc.

(Possibly you mentioned doing that in another PR later, if so, nvm!)

MacDeluca replied:
Updated.

NickPhura commented on api/src/services/import-services/sampling-periods/import-sample-periods-service.ts:
```diff
@@ -0,0 +1,144 @@
+
+    await samplePeriodService.insertSamplePeriods(this.surveyId, samplePeriods);
+
+    defaultLog.debug({ label: 'import sample periods', samplePeriods });
```
Re logging in general: We don't have a strict pattern established per se, but I think it might be worth including the name of the function in the log:

```
defaultLog.debug({ label: 'importCSVWorksheet', message: 'import sample periods', samplePeriods });
```

That combined with the label that gets set at the file level when you call getLogger helps narrow down where the log message came from more precisely.

NickPhura replied:
I was going to make the label and message fields required as part of that previous logging ticket, but there are some logs in the app that dont have them, and I didnt want to make a giant PR

MacDeluca replied:
Updated.

NickPhura commented:
Do we need to add a validation function for overlapping dates?

I tried a file with bad dates, and the SQL fails as expected, but it just results in a generic request error on the frontend, as opposed to a more human friendly validation error.

NickPhura commented:
I tried file with valid dates, and they get imported successfully. However, it looks like the dates are getting parsed incorrectly.

I had start date `01/08/2007` and end date `31/12/2007`
The end date worked as expected, but the start date was inserted as `January 8th`, when I was expecting `August 1st`.

When I changed excel to `yyyy-mm-dd` then it did work as expected.

The file I was testing with:
[Bulk.Periods.csv](https://github.com/user-attachments/files/18711527/Bulk.Periods.csv)

MacDeluca commented:
> I tried file with valid dates, and they get imported successfully. However, it looks like the dates are getting parsed incorrectly.
> 
> I had start date `01/08/2007` and end date `31/12/2007` The end date worked as expected, but the start date was inserted as `January 8th`, when I was expecting `August 1st`.
> 
> When I changed excel to `yyyy-mm-dd` then it did work as expected.
> 
> The file I was testing with: [Bulk.Periods.csv](https://github.com/user-attachments/files/18711527/Bulk.Periods.csv)

I think this was from not handling single digits (1/8/2024) for days / months. I updated the `formatDateString` function to handle those cases. I also updated the tests to be slightly more clear about which dates it needs to validate correctly.

MacDeluca commented:
> Do we need to add a validation function for overlapping dates?
> 
> I tried a file with bad dates, and the SQL fails as expected, but it just results in a generic request error on the frontend, as opposed to a more human friendly validation error.

Not yet, Macgregor said to prioritize getting this in first.

NickPhura commented on api/src/utils/csv-utils/row-validators/start-end-date-order-row-validator.ts:
```diff
@@ -0,0 +1,64 @@
+    let startTimeStamp = utils.getCellValue(staticHeaders.startDate, params.row) as string;
+    let endTimeStamp = utils.getCellValue(staticHeaders.endDate, params.row) as string;
+
+    if (staticHeaders.startTime) {
```
It looks like `staticHeaders.startTime` is always truthy, and so the `startTimeStamp` is being created as `'2007-08-11 undefined'`. Same for the `endTimeStamp`

Should this `if` be checking the cell value instead?

NickPhura replied:
In this case, I was testing with a file that had no cell values in either time column.

MacDeluca replied:
Ooops brain dont work gud anymore. 

Fixed :)

MacDeluca committed:
fix: renamed static headers + added additional tests for date validation

MacDeluca committed:
chore: added tests for start end date row validator

MacDeluca committed:
fix: removed id from insert error

NickPhura reviewed (dismissed):
🧪 ⌛

MacDeluca committed:
fix: dropped only

MacDeluca committed:
ignore-skip

MacDeluca committed:
Merge branch 'dev' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-664-sampling-periods-csv

MacDeluca committed:
fix: dropped SAMPLING_SITE

NickPhura reviewed (approved):
🦙
